smoketest/graphqlnotifiedalerts_test.go

```diff
@@ -0,0 +1,153 @@
+package smoketest
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/target/goalert/smoketest/harness"
+)
+
+func TestNotifiedAlerts(t *testing.T) {
+	t.Parallel()
+
+	sql := `
+	insert into users (id, name, email, role) 
+	values ({{uuid "user"}}, 'bob', 'joe', 'user');
+
+	insert into user_contact_methods (id, user_id, name, type, value) 
+	values ({{uuid "cm1"}}, {{uuid "user"}}, 'personal', 'SMS', {{phone "1"}});
+
+	insert into user_notification_rules (user_id, contact_method_id, delay_minutes) 
+	values ({{uuid "user"}}, {{uuid "cm1"}}, 0);
+
+	insert into escalation_policies (id, name) 
+	values
+		({{uuid "eid"}}, 'esc policy'),
+		({{uuid "eid2"}}, 'esc policy 2');
+
+	insert into escalation_policy_steps (id, escalation_policy_id) 
+	values ({{uuid "esid"}}, {{uuid "eid"}});
+ 
+
+	insert into services (id, escalation_policy_id, name) 
+	values
+		({{uuid "sid"}}, {{uuid "eid"}}, 'service'),
+		({{uuid "sid2"}}, {{uuid "eid2"}}, 'service 2');
+	`
+
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	doQL := func(t *testing.T, h *harness.Harness, query string, res interface{}) {
+		t.Helper()
+		g := h.GraphQLQuery2(query)
+		for _, err := range g.Errors {
+			t.Error("GraphQL Error:", err.Message)
+		}
+		if len(g.Errors) > 0 {
+			t.Fatal("errors returned from GraphQL")
+		}
+		t.Log("Response:", string(g.Data))
+		if res == nil {
+			return
+		}
+		err := json.Unmarshal(g.Data, &res)
+		if err != nil {
+			t.Fatal("failed to parse response:", err)
+		}
+	}
+
+	doQL(t, h, fmt.Sprintf(`
+		mutation{
+			setFavorite(input:{
+				target:{
+					id: "%s"
+					type: service
+				}
+				favorite: true
+			})
+		}
+	`, h.UUID("sid2")), nil)
+
+	var s struct {
+		Service struct {
+			IsFavorite bool
+		}
+	}
+
+	doQL(t, h, fmt.Sprintf(`
+		query {
+			service(id: "%s") { 
+				isFavorite 
+			}	
+		}
+	`, h.UUID("sid2")), &s)
+
+	if s.Service.IsFavorite != true {
+		t.Fatalf("ERROR: ServiceID %s IsUserFavorite=%t; want true", h.UUID("sid2"), s.Service.IsFavorite)
+	}
+
+	h.CreateAlert(h.UUID("sid"), "alert1")
+	h.CreateAlert(h.UUID("sid2"), "alert2")
+
+	type Alerts struct {
+		Alerts struct {
+			Nodes []struct {
+				ID string
+			}
+		}
+	}
+
+	var alerts1, alerts2, alerts3 Alerts
+
+	// output: 1 alert (the favorited one)
+	doQL(t, h, `query {
+		alerts(input: {
+			includeNotified: false
+			favoritesOnly: true
+		}) {
+			nodes {
+				id
+			}
+		}
+	}`, &alerts1)
+
+	if len(alerts1.Alerts.Nodes) != 1 {
+		t.Errorf("got %d alerts; want 1", len(alerts1.Alerts.Nodes))
+	}
+
+	// output: 1 alert (alert from favorited and alert from notified are merged together)
+	doQL(t, h, `query {
+			alerts(input: {
+				includeNotified: true
+				favoritesOnly: true
+			}) {
+				nodes {
+					id
+				}
+			}
+		}`, &alerts2)
+
+	if len(alerts2.Alerts.Nodes) != 1 {
+		t.Errorf("got %d alerts; want 1", len(alerts2.Alerts.Nodes))
+	}
```

Comment:
I think we may need to think a little more on this test case, as the UI _does_ state that it displays both sets of alerts together:

![image](https://user-images.githubusercontent.com/11381794/88714589-a3f2b500-d0e2-11ea-98b2-dad3bc8973b7.png)

and looking at the code for what is set, it appears that both `includeNotified` and `favoritesOnly` are set to `true`:

```
// default to favorites only, unless viewing alerts from a service's page
favoritesOnly: !props.serviceID && !allServices,
includeNotified: !props.serviceID, // keep service list alerts specific to that service
```

I think we will still have to use the user that is created in the harness (`bcefacc0-4764-012d-7bfb-002500d5decb` from `harness/graphql.go`), but I'm not sure why we were seeing foreign constraint errors when trying to add a contact method in the init SQL logic. Perhaps the user isn't created yet when the harness runs the initSql query? @mastercactapus I would appreciate your insight as well!

Reply:
That is correct, the user get's created after `initSQL` -- initSQL runs at a particular migration level (specified in the `NewHarness` func) but creating a new user uses the User store method, which has to run after all the other migrations have been applied.

It would be possible to make the user ID available earlier but a little complicated (we'd have to add the user at an early migration, then go to init level, then finish, then edit/update the user to match what we need).

What might be better, is to make a gql "client" to replace all the `doQL` calls and allow specifying the user id

That way we can setup initSQL and then use THAT user to make the calls

Reply:
Thanks for the feedback. I am not sure how to setup a gql "client." Is there an existing similar file where this has already been done I could look at?

Reply:
I think the easiest thing, for now, is a graphQL method on the harness where the user ID can be passed in. Maybe in its own PR though. Then the existing one can just call using a fixed user ID and this can specify it's own